packages/eslint-plugin/src/rules/non-nullable-type-assertion-style.ts

```diff
@@ -0,0 +1,104 @@
+import { TSESTree } from '@typescript-eslint/experimental-utils';
+import * as tsutils from 'tsutils';
+import * as ts from 'typescript';
+
+import * as util from '../util';
+
+export default util.createRule({
+  name: 'non-nullable-type-assertion-style',
+  meta: {
+    docs: {
+      category: 'Best Practices',
+      description:
+        'Prefers a non-null assertion over explicit type cast when possible',
+      recommended: false,
+      requiresTypeChecking: true,
+      suggestion: true,
+    },
+    fixable: 'code',
+    messages: {
+      preferNonNullAssertion:
+        'Use a ! assertion to more succintly remove null and undefined from the type.',
+    },
+    schema: [],
+    type: 'suggestion',
+  },
+  defaultOptions: [],
+
+  create(context) {
+    const parserServices = util.getParserServices(context);
+    const checker = parserServices.program.getTypeChecker();
+    const sourceCode = context.getSourceCode();
+
+    const getTypesIfNotLoose = (node: TSESTree.Node): ts.Type[] | undefined => {
+      const type = checker.getTypeAtLocation(
+        parserServices.esTreeNodeToTSNodeMap.get(node),
+      );
+
+      return tsutils.isTypeFlagSet(
+        type,
+        ts.TypeFlags.Any | ts.TypeFlags.Unknown,
+      )
+        ? undefined
+        : tsutils.unionTypeParts(type);
```

Comment:
nitpick - IMO it's clearer without a ternary

```suggestion
      if (
        tsutils.isTypeFlagSet(type, ts.TypeFlags.Any | ts.TypeFlags.Unknown)
      ) {
        return undefined;
      }
      return tsutils.unionTypeParts(type);
```